modules/s2i/core/bash/v9/artifacts/opt/jboss/container/s2i/core/s2i-core

```diff
@@ -0,0 +1,151 @@
+    chmod -R ug+rwX "${S2I_ARTIFACTS_DIR}"
+  fi
+  if [ -d "${S2I_SOURCE_DIR}" ]; then
+    chmod -R ug+rwX "${S2I_SOURCE_DIR}"
+  fi
+}
+
+function s2i_core_env_init() {
+  # initialize core s2i environmet variables
+  s2i_core_env_init_hook
+  S2I_DESTINATION_DIR="${S2I_DESTINATION_DIR:-/tmp}"
+  S2I_ARTIFACTS_DIR="${S2I_DESTINATION_DIR}/artifacts"
+  S2I_SOURCE_DIR="${S2I_DESTINATION_DIR}/src"
+  S2I_SOURCE_CONFIGURATION_DIR="${S2I_SOURCE_CONFIGURATION_DIR:-configuration}"
+  S2I_SOURCE_DATA_DIR="${S2I_SOURCE_DATA_DIR:-${APP_DATADIR:-data}}"
+  S2I_SOURCE_DEPLOYMENTS_DIR="${S2I_SOURCE_DEPLOYMENTS_DIR:-deployments}"
+  S2I_TARGET_DEPLOYMENTS_DIR="${S2I_TARGET_DEPLOYMENTS_DIR:-${DEPLOYMENTS_DIR:-/deployments}}"
+  S2I_TARGET_CONFIGURATION_DIR="${S2I_TARGET_CONFIGURATION_DIR:-${S2I_TARGET_DEPLOYMENTS_DIR}}"
+  S2I_TARGET_DATA_DIR="${S2I_TARGET_DATA_DIR:-${JAVA_DATA_DIR:-${DATA_DIR:-${S2I_TARGET_DEPLOYMENTS_DIR}}}}"
+  S2I_IMAGE_SOURCE_MOUNTS="${S2I_IMAGE_SOURCE_MOUNTS:-${CUSTOM_INSTALL_DIRECTORIES}}"
+  S2I_ENABLE_INCREMENTAL_BUILDS="${S2I_ENABLE_INCREMENTAL_BUILDS:-true}"
+  S2I_DELETE_SOURCE="${S2I_DELETE_SOURCE:-true}"
+}
+
+# extensions may override this method to initialize environment variables
+# to suit their own needs.
+function s2i_core_env_init_hook() {
+  : 
+}
+
+# copy configuration files
+# $1 - the base directory to which $S2I_SOURCE_CONFIGURATION_DIR is appended
+function s2i_core_copy_configuration() {
+  if [ -d "${1}/${S2I_SOURCE_CONFIGURATION_DIR}" ]; then
+    if [ -z "${S2I_TARGET_CONFIGURATION_DIR}" ]; then
+      log_warning "Unable to copy configuration files.  No target directory specified for S2I_TARGET_CONFIGURATION_DIR"
+    else
+      if [ ! -d "${S2I_TARGET_CONFIGURATION_DIR}" ]; then
+        log_info "S2I_TARGET_CONFIGURATION_DIR does not exist, creating ${S2I_TARGET_CONFIGURATION_DIR}"
+        mkdir -pm 775 "${S2I_TARGET_CONFIGURATION_DIR}"
+      fi
+      log_info "Copying configuration from $(realpath --relative-to ${S2I_SOURCE_DIR} ${1}/${S2I_SOURCE_CONFIGURATION_DIR}) to ${S2I_TARGET_CONFIGURATION_DIR}..."
+      rsync -rl --out-format='%n' "${1}/${S2I_SOURCE_CONFIGURATION_DIR}"/ "${S2I_TARGET_CONFIGURATION_DIR}"
+    fi
+  fi 
+}
+
+# copy data files
+# $1 - the base directory to which $S2I_SOURCE_DATA_DIR is appended
+function s2i_core_copy_data() {
+  if [ -d "${1}/${S2I_SOURCE_DATA_DIR}" ]; then
+    if [ -z "${S2I_TARGET_DATA_DIR}" ]; then
+      log_warning "Unable to copy data files.  No target directory specified for S2I_TARGET_DATA_DIR"
+    else
+      if [ ! -d "${S2I_TARGET_DATA_DIR}" ]; then
+        log_info "S2I_TARGET_DATA_DIR does not exist, creating ${S2I_TARGET_DATA_DIR}"
+        mkdir -pm 775 "${S2I_TARGET_DATA_DIR}"
+      fi
+      log_info "Copying app data from $(realpath --relative-to ${S2I_SOURCE_DIR} ${1}/${S2I_SOURCE_DATA_DIR}) to ${S2I_TARGET_DATA_DIR}..."
+      rsync -rl --out-format='%n' "${1}/${S2I_SOURCE_DATA_DIR}"/ "${S2I_TARGET_DATA_DIR}"
```

Comment:
In passing, perhaps reconsider whether rsync is needed in the OpenJDK container images. It's not clear what the use-case for this is, though.